packages/tpetra/core/src/Tpetra_Details_DistributorActor.hpp

```diff
@@ -0,0 +1,796 @@
+
+  template <class ExpView, class ImpView>
+  void doPostsAndWaits(const DistributorPlan& plan,
+                       const ExpView &exports,
+                       size_t numPackets,
+                       const ImpView &imports);
+
+  template <class ExpView, class ImpView>
+  void doPostsAndWaits(const DistributorPlan& plan,
+                       const ExpView &exports,
+                       const Teuchos::ArrayView<const size_t>& numExportPacketsPerLID,
+                       const ImpView &imports,
+                       const Teuchos::ArrayView<const size_t>& numImportPacketsPerLID);
+
+  template <class ExpView, class ImpView>
+  void doPosts(const DistributorPlan& plan,
+               const ExpView& exports,
+               size_t numPackets,
+               const ImpView& imports);
+
+  template <class ExpView, class ImpView>
+  void doPosts(const DistributorPlan& plan,
+               const ExpView &exports,
+               const Teuchos::ArrayView<const size_t>& numExportPacketsPerLID,
+               const ImpView &imports,
+               const Teuchos::ArrayView<const size_t>& numImportPacketsPerLID);
+
+  void doWaits(const DistributorPlan& plan);
+
+private:
+  Teuchos::Array<Teuchos::RCP<Teuchos::CommRequest<int>>> requests_;
+
+#ifdef HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_;
+  Teuchos::RCP<Teuchos::Time> timer_doWaits_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_recvs_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_recvs_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_barrier_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_barrier_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_sends_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_sends_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_sends_slow_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_sends_slow_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts3KV_sends_fast_;
+  Teuchos::RCP<Teuchos::Time> timer_doPosts4KV_sends_fast_;
+
+  //! Make the instance's timers.  (Call only in constructor.)
+  void makeTimers();
+#endif // HAVE_TPETRA_DISTRIBUTOR_TIMINGS
+};
+
+template <class ExpView, class ImpView>
+void DistributorActor::doPostsAndWaits(const DistributorPlan& plan,
+                                       const ExpView& exports,
+                                       size_t numPackets,
+                                       const ImpView& imports)
+{
+  static_assert(areKokkosViews<ExpView, ImpView>,
+      "Data arrays for DistributorActor::doPostsAndWaits must be Kokkos::Views");
+  doPosts(plan, exports, numPackets, imports);
+  doWaits(plan);
+}
+
+template <class ExpView, class ImpView>
+void DistributorActor::doPostsAndWaits(const DistributorPlan& plan,
+                                       const ExpView& exports,
+                                       const Teuchos::ArrayView<const size_t>& numExportPacketsPerLID,
```

Comment:
Eventually (not this PR), should all arguments be Kokkos views, or is mix of Kokkos views and Teuchos::ArrayView OK?

Reply:
good point.  perhaps we can discuss at a team meeting in the near future.  I'm not really sure what to think right now.